Mathlib/Geometry/Manifold/Orientable.lean

```diff
@@ -0,0 +1,192 @@
+  admits an orientation.
+- Define orientation preserving and reserving maps between manifolds.
+
+## Implementation notes
+
+The current definitions work for differentiable manifolds. For topological manifolds, orientability
+can be defined using *local* orientations (which mathlib cannot state yet, as there is no e.g.
+singular homology). In the future, it would be nice to generalise these definitions to allow for
+topological manifolds also, and relate them to the current definition.
+
+-/
+
+variable {H : Type*} [NormedAddCommGroup H] [NormedSpace ℝ H]
+
+section OrientationPreserving
+
+/--
+A map between normed spaces is orientation-preserving on a given set if it is differentiable and the
+determinant of its Jacobian is strictly positive on that set.
+-/
+def OrientationPreserving (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s, 0 < (fderiv ℝ f x).det
+
+/--
+A map between normed spaces is orientation-reversing on a given set if it is differentiable and the
+determinant of its Jacobian is strictly negative on that set.
+-/
+def OrientationReversing (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s, (fderiv ℝ f x).det < 0
+
+lemma orientationPreserving_of_zero_dim (f : H → H) (s : Set H) (h : Module.rank ℝ H = 0) :
+    OrientationPreserving f s :=
+  fun _ _ ↦
+  have det_eq_one : (fderiv ℝ f _).det = 1 := by
+    have b : Basis (Fin 0) ℝ H := Basis.ofEquivFun (finDimVectorspaceEquiv 0 h)
+    rw [ContinuousLinearMap.det, ← (fderiv ℝ f _).det_toMatrix b]
+    exact Matrix.det_fin_zero
+  det_eq_one ▸ Real.zero_lt_one
+
+lemma OrientationPreserving.differentiableAt [FiniteDimensional ℝ H] {f : H → H} {s : Set H}
+    (h : OrientationPreserving f s) {x : H} (hs : x ∈ s) : DifferentiableAt ℝ f x := by
+  cases subsingleton_or_nontrivial H
+  · apply DifferentiableWithinAt.differentiableAt
+    apply Set.Subsingleton.differentiableOn
+    exact s.subsingleton_of_subsingleton
+    exact hs
+    exact mem_nhds_discrete.mpr hs
+  · unfold OrientationPreserving at h
+    contrapose! h
+    use x, hs
+    rw [fderiv_zero_of_not_differentiableAt h, ContinuousLinearMap.det]
+    simp [ne_of_gt FiniteDimensional.finrank_pos]
+
+lemma OrientationReversing.differentiableAt {f : H → H} {s : Set H} (h : OrientationReversing f s)
+    {x : H} (hs : x ∈ s) : DifferentiableAt ℝ f x := by
+  unfold OrientationReversing at h
+  contrapose! h
+  use x, hs
+  rw [fderiv_zero_of_not_differentiableAt h, ContinuousLinearMap.det]
+  simp [ne_of_gt FiniteDimensional.finrank_pos]
+
+lemma orientationPreserving_id (s : Set H) : OrientationPreserving id s := by
+  intro
+  simp [ContinuousLinearMap.det]
+
+lemma orientationPreserving_comp [FiniteDimensional ℝ H] {f g : H → H} {u v : Set H}
+    (hf : OrientationPreserving f u) (hg : OrientationPreserving g v) :
+    OrientationPreserving (g ∘ f) (u ∩ f ⁻¹' v) := by
+  intro x ⟨hxu, hxv⟩
+  rw [fderiv.comp x (hg.differentiableAt hxv) (hf.differentiableAt hxu)]
```

Comment:
Here is a golf of the last three lines ("unfold" to "exact").
- avoiding `unfold` in favour of `rw`
- in fact, `simp` can also be used (which makes this even shorter)
- in the last step, `simp` followed by `exact` can often be converged to `simpa [...] using`: I think that is more readable here.
Overall, I'd suggest
```
simpa [ContinuousLinearMap.det] using mul_pos (hg (f x) hxv) (hf x hxu)
```

Please make the same change to the other three lemmas below.